File: arena-core/src/test/java/arena/core/ReplayTest.java
package arena.core;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class ReplayTest {

    private static BotFunction avoid() {
        return view -> {
            for (Direction d : new Direction[]{
                    Direction.RIGHT, Direction.DOWN, Direction.LEFT, Direction.UP}) {
                if (!view.isDeadly(d)) return d;
            }
            return view.myDir();
        };
    }

    @Test
    void moves는_턴당_2문자다() {
        Replay r = Match.play("a", avoid(), "b", avoid(), 5, 30, 30);
        assertEquals(r.result().turns() * 2, r.moves().length());
    }

    @Test
    void moves는_UDLR만_담는다() {
        Replay r = Match.play("a", avoid(), "b", avoid(), 5, 30, 30);
        assertTrue(r.moves().matches("[UDLR]+"), "예상 못한 문자: " + r.moves());
    }

    @Test
    void 같은_경기는_같은_해시를_낸다() {
        Replay a = Match.play("a", avoid(), "b", avoid(), 5, 30, 30);
        Replay b = Match.play("a", avoid(), "b", avoid(), 5, 30, 30);
        assertEquals(a.hash(), b.hash());
        assertTrue(a.hash().startsWith("sha256:"));
    }

    @Test
    void 다른_시드는_다른_해시를_낸다() {
        Replay a = Match.play("a", avoid(), "b", avoid(), 5, 30, 30);
        Replay b = Match.play("a", avoid(), "b", avoid(), 6, 30, 30);
        assertNotEquals(a.hash(), b.hash());
    }

    @Test
    void 리플레이는_시작_배치를_기록한다() {
        StartPositions sp = StartPositions.of(5, 30, 30);
        Replay r = Match.play("alpha", avoid(), "beta", avoid(), 5, 30, 30);

        assertEquals("alpha", r.bot0Id());
        assertEquals("beta", r.bot1Id());
        assertEquals(sp.p0(), r.start0());
        assertEquals(sp.d0(), r.dir0());
        assertEquals(sp.p1(), r.start1());
        assertEquals(sp.d1(), r.dir1());
        assertEquals(5L, r.seed());
    }

    @Test
    void 짧은_경기의_리플레이는_1KB_미만이다() {
        Replay r = Match.play("a", avoid(), "b", avoid(), 5, 30, 30);
        // moves가 전체 크기를 지배한다. 187턴이면 374바이트.
        assertTrue(r.moves().length() < 1800,
                "moves가 " + r.moves().length() + "자나 된다");
    }
}

File: web/src/test/replay.test.ts
import { describe, it, expect } from 'vitest';
import { decodeReplay, owner } from '../lib/replay';
import type { Replay } from '../lib/schema';

/** 첫 턴에 봇0이 뒤로 가고 봇1은 직진하는 최소 경기. */
const 후진_경기: Replay = {
  schema: 1, matchId: 'test-back', width: 30, height: 30, seed: 1, swapped: false,
  bot0Id: 'a', start0: { x: 10, y: 10 }, dir0: 'RIGHT',
  bot1Id: 'b', start1: { x: 20, y: 20 }, dir1: 'LEFT',
  moves: 'LL',
  result: { winner: 1, turns: 1, reason: 'P0_HIT_OWN_WALL' },
  hash: 'sha256:테스트픽스처',
};

describe('시작 격자', () => {
  it('시작 칸 뒤 칸도 벽이다 — 첫 턴 후진이 자기 벽 충돌로 죽는다', () => {
    // 봇0은 RIGHT로 시작하므로 (9,10)이 처음부터 벽이다. LEFT를 내면
    // 거기 박는다. 뒤 칸을 안 잡으면 이 경기는 죽지 않고 계속된다.
    const decoded = decodeReplay(후진_경기);
    expect(decoded.turnCount).toBe(1);
    expect(decoded.winner).toBe(1);
    expect(decoded.reason).toBe('P0_HIT_OWN_WALL');
  });

  it('턴 1의 격자에 벽이 정확히 4칸 있다', () => {
    const decoded = decodeReplay(후진_경기);
    let walls = 0;
    for (let x = 0; x < 30; x++) {
      for (let y = 0; y < 30; y++) if (owner(decoded, 1, x, y) !== null) walls++;
    }
    expect(walls).toBe(4);
  });
});

describe('동시 판정', () => {
  it('같은 칸에 동시 진입하면 무승부다', () => {
    const 정면충돌: Replay = {
      schema: 1, matchId: 'test-headon', width: 30, height: 30, seed: 1, swapped: false,
      bot0Id: 'a', start0: { x: 10, y: 10 }, dir0: 'RIGHT',
      bot1Id: 'b', start1: { x: 12, y: 10 }, dir1: 'LEFT',
      moves: 'RL',
      result: { winner: -1, turns: 1, reason: 'HEAD_ON_COLLISION' },
      hash: 'sha256:테스트픽스처',
    };
    const decoded = decodeReplay(정면충돌);
    expect(decoded.winner).toBe(-1);
    expect(decoded.reason).toBe('HEAD_ON_COLLISION');
  });

  it('죽은 봇의 머리는 벽이 되지 않는다', () => {
    // 봇0만 죽는 턴에 확정되는 벽은 봇1의 머리 하나뿐이다 (스펙 §7.1).
    const decoded = decodeReplay(후진_경기);
    const last = decoded.turns[decoded.turns.length - 1];
    expect(last.claimed[0]).toBeNull();
    expect(last.claimed[1]).not.toBeNull();
  });
});
